### PR TITLE
[WIP] Rework Verilog Backend

### PR DESCRIPTION
**Pull request by hofstee**
The old Verilog backend worked primarily by munging and reformatting strings. This branch reworks it to generate a Verilog AST that is dumped as a string instead.

### CODE REVIEW

**Comment by hofstee** [comment]
@leonardt I don't think this is my fault (no pun intended).

**Comment by leonardt** [comment]
It's likely failing because this checks out magma master but I believe your changes depend on https://github.com/phanrahan/magma/pull/303

Will rerun the build after the magma PR has been merged

File: silica/compile.py
Comment (leonardt) [inline silica/compile.py]: Minor style comment, I know this is a tedious change, but I prefer if we use `module.func` rather than `import *`. Mainly, this interacts better with tooling for things like autocompletion, also makes it more explicit where things are coming from in the code.

We could do
```
import silica.ast_utils as ast_utils
```

then prepend each function call with `ast_utils`, e.g. `ast_utils.is_name(...)`.

We could also use some shorthand to make it easier to type, maybe `import silica.ast_utils as au`? `au` is not super clear through

File: silica/compile.py
Comment (leonardt) [inline silica/compile.py:186]: We should be able to get rid of all this old `verilog_source` related code, correct?

File: silica/compile.py
Comment (leonardt) [inline silica/compile.py:208]: Might be nice to factor these out into separate functions, I can take care of that, just noting it here.

File: silica/verilog.py
Comment (leonardt) [inline silica/verilog.py:108]: Perhaps wrap these functions in a class which is constructed with module, so all these methods can refer to the instance attribute `self.module`?

File: silica/verilog.py
Comment (leonardt) [inline silica/verilog.py:230]: It looks like this may have been done?

File: silica/verilog.py
Comment (leonardt) [inline silica/verilog.py:224]: We can hold off on this for now, since the SSA backend is the intended one to be used.

**Review by leonardt — CHANGES_REQUESTED** [review]
Looks good, just minor style comments that I can help resolve. Major thing is I think we can get rid of the old verilog_source related code where possible

File: silica/compile.py
Comment (hofstee) [reply silica/compile.py:186]: I'm leaving it in for now because some of the old code paths depend on variables that were created and I don't have full coverage of the old paths yet, just enough to pass the existing tests. It helps a lot to be able to dump both strings at any point during codegen and double check the output is as intended. 

Once I add those remaining cases and refactor a bit I'll delete this stuff.

File: silica/verilog.py
Comment (hofstee) [reply silica/verilog.py:108]: Yeah, I have a few ideas of how to refactor this and what other helper functions would be useful. I'm definitely making a context class with the module where all of these will be members.

File: silica/compile.py
Comment (leonardt) [reply silica/compile.py:186]: Got it, let's not remove it for now then.

**Review by leonardt — APPROVED** [review]
Okay, I'm going to merge this in, feel free to continue to work on this (in this branch or a new one), but I'd like to get this into the mainline now so we don't run into integration issues.

**Comment by hofstee** [comment]
Sounds good, go ahead. I'll work on the other issues and merge them in as I complete them.

**Pull request merged**
